include/IECore/MeshAlgo.h

```diff
@@ -0,0 +1,62 @@
+//////////////////////////////////////////////////////////////////////////
+//
+//  Copyright (c) 2017, Image Engine Design Inc. All rights reserved.
+//
+//  Redistribution and use in source and binary forms, with or without
+//  modification, are permitted provided that the following conditions are
+//  met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+//     * Neither the name of Image Engine Design nor the names of any
+//       other contributors to this software may be used to endorse or
+//       promote products derived from this software without specific prior
+//       written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef IECORE_MESHALGO_H
+#define IECORE_MESHALGO_H
+
+#include <vector>
+#include <utility>
+
+namespace IECore
+{
+
+namespace MeshAlgo
+{
+
+template<typename T>
+std::pair<typename T::Ptr, typename T::Ptr> calculateTangents( const std::vector<int> &vertsPerFace,
```

Comment:
A few things for discussion :

- Have we _ever_ used a mesh with V3dVectorData for "P"? I suspect not, and none of the renderers support doubles for vertex data. I'd advocate we drop the templating and just use V3fVectorData. We're inconsistent with where we do the templated-dispatch-for-multiple-types and where we don't in Cortex, and I'd like to take this opportunity to drop some bloat and settle onto a floats-only standard.
- I'd suggest a function signature taking advantage of the bundling of related data that the MeshPrimitive provides, and future proofing us for Cortex 10 :
    - Something like `pair<PrimitiveVariable, PrimitiveVariable> calculateUVTangents( const MeshPrimitive *mesh, const std::string &uvSet )`
    - I can't imagine using this with anything except MeshPrimitive, so why give me the burden of passing all the arguments, and the chance of passing something that is mismatched? We could introduce an overload if we really need it (but see below for why I don't think it's wise).
    - In Cortex 10 we're going to start storing UVs as a single V2fData, so we need a signature that doesn't assume the current horrible split storage we have.
    - In Cortex 10 we're going to ditch the awful uvIndices hack and introduce optional explicit indexing into PrimitiveVariable, so we need a signature that will cope with that.
    - It's possible that the UV set will have Vertex rather than FaceVarying data, and that the result should therefore be Vertex too. So we should have a signature that allows for that by returning a PrimitiveVariable, even if we don't implement it immediately.